src/sentry/static/sentry/app/views/performance/realUserMonitoring/utils.tsx

```diff
@@ -0,0 +1,158 @@
+import {Query} from 'history';
+import {ECharts} from 'echarts';
+
+import {HistogramData, Rectangle, Point} from './types';
+
+export function generateRealUserMonitoringRoute({orgSlug}: {orgSlug: String}): string {
+  return `/organizations/${orgSlug}/performance/summary/rum/`;
+}
+
+export function realUserMonitoringRouteWithQuery({
+  orgSlug,
+  transaction,
+  projectID,
+  query,
+}: {
+  orgSlug: string;
+  transaction: string;
+  query: Query;
+  projectID?: string | string[];
+}) {
+  const pathname = generateRealUserMonitoringRoute({
+    orgSlug,
+  });
+
+  return {
+    pathname,
+    query: {
+      transaction,
+      project: projectID,
+      environment: query.environment,
+      statsPeriod: query.statsPeriod,
+      start: query.start,
+      end: query.end,
+      query: query.query,
+    },
+  };
+}
+
+/**
+ * Given a value on the x-axis, return the index of the nearest bucket or null
+ * if it cannot be found.
+ *
+ * A bucket contains a range a values, and nearest is defined by the distance
+ * of the point to the mid point of the bucket.
+ */
+export function findNearestBucketIndex(
+  chartData: HistogramData[],
+  bucketWidth: number,
+  xAxis: number
+): number | null {
+  // it's possible that the data is not available yet or the x axis is out of range
+  if (
+    !chartData.length ||
+    xAxis < chartData[0].histogram ||
+    xAxis > chartData[chartData.length - 1].histogram
+  ) {
+    return null;
+  }
+
+  const halfBucketWidth = bucketWidth / 2;
+
+  // binary search for the index
```

Comment:
I'm not really sure what the data looks like so grain of salt here, but if you are iterating over a monotonic list you can just iterate through it once and build the buckets / put things into buckets, instead of doing a binary search on each value. You just increment the bucket index as you check the condition for each value.

Reply:
The intent of this method is to find the index of the bin that the measurement belongs in. Since the list of bins is sorted, I figured a binary search would be just a tad bit better than a linear search (there's 100 bins). Though now that i think about it, I might be able to do this in `O(1)` using something similar to the `floor` calculations used in the backend.

Reply:
Addressed in 33a7282